decision-records/2024-08-29--dashboard-uid-collision.md

```diff
@@ -0,0 +1,134 @@
+# Dashboard uid collisions in Grafana
+**Date:** 2024-08-29<br/>
+**Authors:** @lucabello @sed-i
+
+
+## Context and Problem Statement
+Dashboard json includes keys such as in the following example:
+```bash
+$ cat metrics-dashboard.json | jq '{"title", "uid", "version", "tags"}'
+{
+  "title": "Loki Operator Overview",
+  "uid": "loki",
+  "version": 1,
+  "tags": [
+    "loki"
+  ]
+}
+```
+
+Grafana [documented](https://grafana.com/docs/grafana/latest/administration/provisioning/#reusable-dashboard-urls) that dashboards should not have the same *title* within a folder, or the same *uid* within
+installations, which causes some problems in different scenarios.
+
+This ADR addresses **uid** collisions.
+
+#### (a) Two charms on different revisions
+
+`charm(rev1, dashboard v1) <--> grafana <--> charm(rev2, dashboard v2)`
+
+Unless manually modified, `dashboard v1` and `dashboard v2` have the same *title* and *uid*, which will cause problems in Grafana.
+
+#### (b) New dashboard created by cloning an existing dashboard, neglecting to modify the UID
+
+A common way to create a new dashboard is to clone and modify dashboard (local or from the grafana marketplace).
+If the UID isn't modified (need to remeber to do it manually), then we'll end up with a UID collision (but the revision would be auto-incremented by the grafana UI, if saved and exported using grafana).
+
+#### (c) Unrelated dashboards that coincidentally happen to have the same UID
+
+This is likely to happen when charm authors manually modify UIDs, assuming their (naive :) modification produced a unique ID (in the ecosystem!).
+
+
+## Requirements
+1. Dashboards must not have the same *uid* within installation.
+2. Dashboard URL must be stable across dashboard updates (e.g. charm upgrades). Do not generate the UID from hashing
+                                                                                                                                                                                                                                                                                                                                                                                                                             dashboard's entire contents, because it would break links to the dashboard.[^DbURL]
+3. UID length must be <= 40 (grafana limit).
+4. Should be able to easily tell from which charm each dashboard came.
+
+## Decision
+(2/P) + (4/R) + (5/P) + (6).
```

Comment:
We came to a decision at this point in the reading but it is not obvious what the equation `(2/P) + (4/R) + (5/P) + (6)` stands for.

Then you continue reading and can have an idea of what it means but is not clear how this decision is applied in a real example.

I would move the Decision to the bottom of the document so we read the alternatives first.

Reply:
I inlined the decisions.
The section is at the top per template.